zcash_client_sqlite/src/wallet/scanning.rs

```diff
@@ -0,0 +1,1107 @@
+            "SELECT subtree_end_height
+            FROM sapling_tree_shards
+            WHERE shard_index = :shard_index",
+        )?;
+
+        // if no notes belonging to the wallet were found, so don't need to extend the scanning
+        // range suggestions to include the associated subtrees, and our bounds are just the
+        // scanned range
+        subtree_bounds
+            .map(|(min_idx, max_idx)| {
+                let range_min = if *min_idx > 0 {
+                    // get the block height of the end of the previous shard
+                    sapling_shard_end_stmt
+                        .query_row(named_params![":shard_index": *min_idx - 1], |row| {
+                            row.get::<_, Option<u32>>(0)
+                                .map(|opt| opt.map(BlockHeight::from))
+                        })
+                        .optional()?
+                        .flatten()
+                } else {
+                    // our lower bound is going to be the Sapling activation height
+                    params.activation_height(NetworkUpgrade::Sapling)
+                };
+
+                // get the block height for the end of the current shard
+                let range_max = sapling_shard_end_stmt
+                    .query_row(named_params![":shard_index": max_idx], |row| {
+                        row.get::<_, Option<u32>>(0)
+                            .map(|opt| opt.map(BlockHeight::from))
+                    })
+                    .optional()?
+                    .flatten();
+
+                Ok::<Range<BlockHeight>, rusqlite::Error>(match (range_min, range_max) {
+                    (Some(start), Some(end)) => Range { start, end },
+                    (Some(start), None) => Range {
+                        start,
+                        end: range.end,
+                    },
+                    (None, Some(end)) => Range {
+                        start: range.start,
+                        end,
+                    },
+                    (None, None) => range.clone(),
+                })
+            })
+            .transpose()
+            .map_err(SqliteClientError::from)
+    }?;
+
+    let query_range = extended_range.clone().unwrap_or_else(|| range.clone());
+
+    let scanned = ScanRange::from_parts(range.clone(), ScanPriority::Scanned);
+    let extensions = if let Some(extended) = extended_range {
+        vec![
+            ScanRange::from_parts(extended.start..range.start, ScanPriority::FoundNote),
+            ScanRange::from_parts(range.end..extended.end, ScanPriority::FoundNote),
+        ]
+    } else {
+        vec![]
+    };
+
+    replace_queue_entries(
+        conn,
+        &query_range,
+        Some(scanned).into_iter().chain(extensions.into_iter()),
+    )?;
+
+    Ok(())
+}
+
+pub(crate) fn update_chain_tip<P: consensus::Parameters>(
+    conn: &rusqlite::Transaction<'_>,
+    params: &P,
+    new_tip: BlockHeight,
+) -> Result<(), SqliteClientError> {
```

Comment:
Bug: `new_tip` is the latest height, and thus an inclusive upper bound. But it is used as an exclusive upper bound throughout this function, meaning that the tip block is never marked for scanning.
```suggestion
    new_tip: BlockHeight,
) -> Result<(), SqliteClientError> {
    // `ScanRange` uses an exclusive upper bound.
    let chain_end = new_tip + 1;

```